README.md
fusionModel
================
Kevin Ummel (<ummel@sas.upenn.edu>)

## Overview

**fusionModel** enables variables unique to a “donor” dataset to be
statistically simulated for (i.e. *fused to*) a “recipient” dataset. The
resulting “fused data” contains all of the recipient *and* donor
variables. The latter are true “synthetic” data – *not* observations
sampled/matched from the donor – and resemble the original donor data in
key respects. fusionModel provides a simple and efficient interface for
general data fusion in *R*. The current release is a beta version.

The package was originally developed to allow statistical integration of
microdata from disparate social surveys. fusionModel is the data fusion
workhorse underpinning the larger fusionACS data platform under
development at the [Socio-Spatial Climate
Collaborative](https://web.sas.upenn.edu/sociospatialclimate/). In this
context, fusionModel is used to fuse variables from a range of social
surveys onto microdata from the American Community Survey, allowing for
analysis and spatial resolution otherwise impossible.

fusionModel can also be used for “pure” data synthesis; i.e. creation of
a wholly synthetic version of a single dataset. This is a specific case
of the more general data fusion problem.

## Methodology

**fusionModel** builds on techniques developed for data synthesis and
statistical disclosure control; e.g. the
[synthpop](https://cran.r-project.org/web/packages/synthpop/index.html)
package ([Nowok, Raab and Dibben
2016](https://doi.org/10.18637%2Fjss.v074.i11)). It uses classification
and regression tree (CART) models ([Breiman et
al. 1984](https://www.routledge.com/Classification-and-Regression-Trees/Breiman-Friedman-Stone-Olshen/p/book/9780412048418);
see [rpart](https://cran.r-project.org/web/packages/rpart/index.html))
to partition donor observations into low-variance nodes. Observations in
a given node are randomly sampled to create simulated values for
recipient observations assigned to the same node, as originally
introduced by [Reiter
(2005)](https://nces.ed.gov/FCSM/pdf/2003FCSM_Reiter.pdf). In the
continuous case, kernel density estimation is used to create a “smooth”
conditional probability distribution for each node. This nonparametric
approach is used to sequentially simulate the fusion variables, allowing
previously-simulated variables to become predictors in subsequent models
(i.e. “chained” models).

The package contains a number of features and innovations designed to
improve performance across intended use cases:

-   Pseudo-optimal ordering of the fusion variables is determined from
    analysis of predictor importance in fully-specified `rpart` models
    fit upfront.

-   LASSO regression is used to “pre-screen” predictor variables prior
    to calling `rpart()`. Predictors for which the LASSO coefficient is
    shrunk to zero are excluded from consideration. This can speed up
    tree-building considerably for larger datasets.

-   A K-means clustering strategy is used to “collapse” the levels of
    unordered factor predictor variables. This allows much faster
    tree-building when categorical variables with many levels are used
    to model unordered factor response variables.

-   Highly-linear or otherwise strictly dependent pairwise relationships
    are automatically detected and preserved in the simulation output.
    Such relationships can be problematic for decision tree simulation
    algorithms.

-   When LASSO pre-screening is enabled, the LASSO results are used to
    determine if a fusion variable should be simulated using a linear
    model rather than a decision tree. This allows highly-linear
    multivariate relationships to be modeled as such explicitly.

-   For continuous and ordered factor data types, the fusion/simulation
    step (optionally) identifies a minimal-change “reshuffling” of
    initial simulated values that induces more realistic rank
    correlations with other variables. *This feature is experimental*.

-   The model-building process is fully parallel on UNIX-like systems.

-   Key parts of the code utilize
    [data.table](https://cran.r-project.org/web/packages/data.table/index.html)
    and
    [biglm](https://cran.r-project.org/web/packages/biglm/index.html)
    functionality for more efficient computation with larger datasets.

-   Fitted `rpart` models are “slimmed” to retain only the information
    absolutely necessary for the data fusion process, leading to reduced
    file size when saved to disk and improved load times.

## Installation

``` r
devtools::install_github("ummel/fusionModel")
library(fusionModel)
```

## Data fusion example

The fusionModel package contains example microdata constructed from the
2015 Residential Energy Consumption Survey (see `?recs` for details and
variable definitions). For real-world use cases, the donor and recipient
input datasets are typically independent and possibly very different in
the number of observations. For illustrative purposes, we will use the
`recs` dataset to create both our “donor” and “recipient” data. This
will also allow us to isolate the performance of fusionModel’s
algorithms.

``` r
# Donor dataset
donor <- recs
dim(donor)
```

    [1] 5686   28

``` r
# Recipient dataset
# Retain a handful of variables we will treat as "predictors" common to both donor and recipient
recipient <- subset(recs, select = c(division, urban_rural, climate, income, age, race))
head(recipient)
```

                division urban_rural                  climate            income age
    1            Pacific       Urban IECC climate zones 3B-4B  $140,000 or more  42
    2 West South Central       Rural IECC climate zones 1A-2A $20,000 - $39,999  60
    3 East South Central       Urban     IECC climate zone 3A $20,000 - $39,999  73
    4 West North Central       Urban     IECC climate zone 4A $40,000 - $59,999  69
    5    Middle Atlantic       Urban     IECC climate zone 5A $40,000 - $59,999  51
    6        New England       Urban IECC climate zones 6A-6B Less than $20,000  33
                                  race
    1                            White
    2                            White
    3                            White
    4 American Indian or Alaska Native
    5                            Asian
    6                            White

The `recipient` dataset contains 6 variables that are shared with
`donor`. These shared “predictor” variables provide a statistical link
between the two datasets. fusionModel exploits the information in these
shared variables.

There are 22 non-shared variables that are unique to `donor`. These are
the variables that will be fused to `recipient`. This includes a mix of
continuous, logical, ordered factor, and unordered factor variables.

``` r
# The variables to be fused
fusion.vars <- setdiff(names(donor), names(recipient))
fusion.vars
```

     [1] "education"      "employment"     "hh_size"        "renter"        
     [5] "home_type"      "year_built"     "square_feet"    "insulation"    
     [9] "heating"        "equipm"         "aircon"         "centralac_age" 
    [13] "televisions"    "disconnect"     "electricity"    "natural_gas"   
    [17] "fuel_oil"       "propane"        "propane_btu"    "propane_expend"
    [21] "use_ng"         "have_ac"       

``` r
# The types of variables to be fused
table(sapply(donor[fusion.vars], vctrs::vec_ptype_abbr))
```


    dbl fct int lgl ord 
      5   5   4   3   5 

We build our fusion model using the `train()` function. The minimal
usage is shown below. See `?train` for additional function arguments and
options. Note that observation weights are ignored here for simplicity
but can be incorporated via the optional `weights` argument.

``` r
fit <- train(data = donor, y = fusion.vars)
```

    22 fusion variables
    6 initial predictor variables
    5686 observations
    Searching for derivative relationships...
    Detected 3 derivative variable(s) that can be omitted from modeling
    Determining order of fusion variables...
    Building fusion models...

The resulting object (`fit`) contains all of the information necessary
to statistically fuse the `fusion.vars` to *any* recipient dataset
containing the necessary shared predictors. Fusion is performed using
the `fuse()` function.

``` r
sim <- fuse(data = recipient, train.object = fit)
```

The output from `fuse()` contains simulated/synthetic values for each of
the `fusion.vars` for each observation in `recipient`. The order of the
columns reflects the order in which the variables were fused. A
pseudo-optimal order is determined automatically within `train()`. Let’s
look at just a few of the simulated variables.

``` r
head(sim[, 1:7])
```

                                          aircon           employment
    1            Central air conditioning system Not employed/retired
    2            Central air conditioning system Not employed/retired
    3 Both a central system and individual units Not employed/retired
    4            Central air conditioning system Not employed/retired
    5            Central air conditioning system   Employed full-time
    6   Individual window/wall or portable units Not employed/retired
                                 heating
    1 Natural gas from underground pipes
    2           Do not use space heating
    3 Natural gas from underground pipes
    4                        Electricity
    5 Natural gas from underground pipes
    6                  Fuel oil/kerosene
                                                                                education
    1                                             Bachelor's degree (for example: BA, BS)
    2                                                          High school diploma or GED
    3                                                          High school diploma or GED
    4                                                  Some college or Associate's degree
    5 Master's, Professional, or Doctorate degree (for example: MA, MS, MBA, MD, JD, PhD)
    6                                                Less than high school diploma or GED
      hh_size   year_built                                    home_type
    1       3 2000 to 2009                 Single-family detached house
    2       1 2000 to 2009                 Single-family detached house
    3       1 2000 to 2009                 Single-family detached house
    4       1 1970 to 1979                 Single-family attached house
    5       3 1980 to 1989                 Single-family detached house
    6       2  Before 1950 Apartment in a building with 5 or more units

**If you run the same code yourself, your results for `sim` *will look
different*.** This is because each call to `fuse()` produces a different
random sampling from the underlying, conditional probability
distributions (see section below on “Generating implicates”).

## Validation

Successful fusion should result in simulated/synthetic variables that
“look like” the donor in key respects. We can perform a series of
comparisons to confirm that this is the case.

The continuous variables in `recs` – like many social survey variables –
can be quite sparse (lots of zeros). Let’s first check that the
proportion of zero values is similar in the donor and simulated data.

              hh_size natural_gas square_feet electricity televisions
    donor           0      0.4193           0           0      0.0239
    simulated       0      0.4193           0           0      0.0241
              propane_expend fuel_oil propane propane_btu
    donor             0.8992   0.9483  0.8992      0.8992
    simulated         0.8996   0.9543  0.8996      0.8996

Comparatively few households use propane or fuel oil, and almost
everyone has a television.

Notice that the proportion of zero values is identical for the “propane”
(gallons consumed), “propane\_btu” (thousand Btu), and “propane\_expend”
(dollars spent) variables. These variables *must* be either all zero or
all non-zero for any given household. Such “structural zeros” are not
uncommon in survey microdata, and they are typically difficult for
tree-based modeling techniques to replicate. fusionModel automatically
detects structural zero patterns and ensures they are correctly
represented in the fusion output.

Now, let’s look at the means of the non-zero values.

              hh_size natural_gas square_feet electricity televisions
    donor      2.5774    576.6752    2081.443    11028.97      2.4195
    simulated  2.6017    589.5476    2112.704    11073.57      2.3975
              propane_expend fuel_oil  propane propane_btu
    donor           672.0280 502.8666 346.7819    31672.67
    simulated       657.8552 489.4481 337.5437    30829.06

Notice that the ratio of mean “propane\_btu” to “propane” is the same
for the donor and simulated datasets (ratio = 91.333). This is as
expected, since the two variables report the same phenomena (propane
consumption) with different units of measurement (gallons vs. Btu). That
is, the relationship is linear. However, decision trees struggle to
capture *strictly* linear relationships. fusionModel automatically

NEWS.md
# fusionModel 0.3.0

* Training and fusion process now utilizes either `rpart()` or `biglm()` models, the latter selected only when highly-predictive (R-squared > 0.99). Relevant changes made to internal function `fitRpart()`.

* `train()` includes detection of pairwise "derivative" variables; i.e. cases where one variable is fully dependent upon another. This has two flavors: either a linear model when Pearson correlation is > 0.99^2 or explicit merge for categorical derivatives. `fuse()` handles simulation accordingly. See `detectDependence.R`.

* Related to above, `train()` attempts to detect "structural zeros" in donor data and `fuse()` preserves that structure in the fusion output.

* `train()` has new arguments to control speed of initial model-fitting and to explicitly pass `cp` and `minbucket` arguments to `rpart()`. Ellipsis argument removed. See `?train`.

* `train()` allows user to set number of cores via new `cores` argument. Defaults to one. Old argument `mc` removed.

* `use.biglm` argument removed from `fuse()`; it is now default when `induce = TRUE`.

* `recs` example data expanded to include more variables.

* README updated.

* Various bug fixes.

# fusionModel 0.2.0

* `train()` is now fully parallel when `mc = TRUE` on UNIX-like systems.

* `lasso` argument added to `train()` enabling fast pre-screening of predictors. See `?train`.

* Fixed bug preventing disabling of `maxcats` argument in `train()`. Now disabled by default (`maxcats = NULL`).

* `induce` argument added to `fuse()` to (optionally) turn off induced rank correlation step.

* [`pbapply`](https://cran.r-project.org/web/packages/pbapply/index.html) package now required for multicore and progress bars.

* [`glmnet`](https://cran.r-project.org/web/packages/glmnet/index.html) package now suggested for LASSO pre-screening.

# fusionModel 0.1.0

* Initial beta.


R/train.R
#' Train a fusion model on donor data
#'
#' @description
#' Train a fusion model on donor data.
#'
#' @param data Data frame. Donor dataset. All categorical variables must be factors and ordered whenever possible.
#' @param y Character vector. The variables to fuse to a recipient dataset. Can include \link[base:regex]{regular expressions}.
#' @param x Character vector. Predictor variables common to donor and eventual recipient. Can include regular expressions. If NULL, all variables other than those in \code{y} and \code{weight} are used. Only one of \code{x} and \code{ignore} can be non-NULL.
#' @param ignore Character vector. Alternative way to specify predictor variables. Can include regular expressions. If non-NULL, all variables other than those in \code{y}, \code{weight}, and \code{ignore} are used. Only one of \code{x} and \code{ignore} can be non-NULL.
#' @param weight Character vector. Name of the observation weights column. If NULL (default), uniform weights are assumed.
#' @param cores Integer. Number of cores used for potential parallel operations. Passed to \code{cl} argument of \code{\link[pbapply]{pblapply}}. Ignored on Windows systems.
#' @param lasso Numeric (0-1) or NULL. Controls extent of predictor variable pre-screening via LASSO regression. If NULL, no pre-screening is performed. Default value \code{lasso = 1} invokes the least-restrictive LASSO. See Details.
#' @param maxcats Positive integer or NULL. Maximum number of levels allowed in an unordered factor predictor variable when the response (fusion) variable is also an unordered factor. Prevents excessive \code{\link[rpart]{rpart}} computation time. A K-means clustering strategy is used to cluster the predictor to no more than \code{maxcats} levels.
#' @param complexity Numeric. Passed to \code{cp} argument of \code{\link[rpart]{rpart.control}} to control complexity of decision trees.
#' @param node.obs Numeric vector of length 2. Minimum number of observations in tree nodes. First number is for numeric response variables; second for categorical. Each is passed to \code{minbucket} argument of \code{\link[rpart]{rpart.control}}.
#' @param initial Numeric vector of length 2. Controls speed/performance of the initial model-fitting routune used to determine fusion order. First number is proportion of \code{data} observations to randomly sample. Second number is the \code{cp} argument passed to \code{\link[rpart]{rpart.control}}. See Details.
#'
#' @details When \code{lasso} is non-NULL, predictor variables are "pre-screened" using LASSO regression via \code{\link[glmnet]{glmnet}} prior to fitting a \code{\link[rpart]{rpart}} model. Predictors with a LASSO coefficient of zero are excluded from consideration. This can speed up tree-fitting considerably when \code{data} is large. Lower values of \code{lasso} are more aggressive at excluding predictors; the LASSO \emph{lambda} is chosen such that the deviance explained is at least \code{lasso}-% of the maximum. To ensure the LASSO step is fast, pre-screening is only used for numeric, logical, and ordered factor response variables (the latter integerized).
#' @details Since determination of the fusion order only makes use of variable importance results from the initial (fully-specified) models, employing a random subset of observations and less complex models (controlled via \code{initial}) can yield a competitive fusion order at less expense.
#'
#' @return A list containing trained model information to be passed to \link{fuse}.
#'
#' @examples
#' donor <- recs
#' recipient <- subset(recs, select = c(division, urban_rural, climate, income, age, race))
#' fusion.vars <- setdiff(names(donor), names(recipient))
#' fit <- train(data = donor, y = fusion.vars)
#' @export

#---------------------

# Manual testing
#
# library(fusionModel)
# source("R/utils.R")
# source("R/fitRpart.R")
# source("R/fitDensity.R")
# source("R/fusionOrder.R")
# source("R/detectDependence.R")
#
# Inputs for ?train example - with some modification for harder test cases
# data <- recs
#
# i <- sample(which(data$propane > 0), size = 100, replace = FALSE)
# data$propane[i] <- -data$propane[i]
# data$propane_btu[i] <- -data$propane_btu[i]
# data$propane_expend[i] <- -data$propane_expend[i]
#
# data <- data %>%
#   mutate(cenac_age = ifelse(centralac_age == "Less than 2 years old", 1, 0),
#          cenac_age = ifelse(centralac_age %in% c("2 to 4 years old", "5 to 9 years old"), 4.5, cenac_age),
#          cenac_age = ifelse(centralac_age %in% c("10 to 14 years old", "15 to 19 years old", "20 years or older"), 17, cenac_age))
#
# recipient <- subset(recs, select = c(division, urban_rural, climate, income, age, race))
# y = setdiff(names(data), names(recipient))
# weight = NULL
# x <- NULL
# ignore = NULL
# maxcats = 10
# cores = 1
# lasso = 1
# complexity = 0.0001
# node.obs = c(50, 10)
# initial <- c(0.5, 0.001)

#---------------------

train <- function(data,
                  y,
                  x = NULL,
                  ignore = NULL,
                  weight = NULL,
                  cores = 1,
                  lasso = 1,
                  maxcats = 10,
                  complexity = 0,
                  node.obs = c(50, 10),
                  initial = c(0.5, 0.001)) {

  stopifnot(exprs = {
    is.data.frame(data)
    !missing(y)
    is.null(maxcats) | (maxcats > 1 & maxcats %% 1 == 0)
    cores > 0 & cores %% 1 == 0
    is.null(lasso) | (lasso > 0 & lasso <= 1)
    complexity >= 0
    length(node.obs) == 2 & all(node.obs > 0)
    is.numeric(initial) & length(initial) == 2
  })

  # Check that no more than one of 'x' or 'ignore' is specified
  if (!is.null(x) & !is.null(ignore)) stop("Only one of 'x' or 'ignore' may be non-NULL")

  #-----

  # Detect fusion variables
  nms <- names(data)
  yvars <- validNames(y, nms, exclude = FALSE)

  # Detect predictor variables
  xvars <- setdiff(nms, c(yvars, weight))
  if (!is.null(ignore)) xvars <- validNames(ignore, xvars, exclude = TRUE)

  # Check validity of variables
  stopifnot(exprs = {
    length(yvars) > 0
    length(xvars) > 0
    #!anyNA(data[yvars])  # Require non-NA values in response values
  })

  #-----

  # Create and/or check observation weights
  # Weights are normalized (mean = 1) to avoid integer overflow issues
  w <- "._wgt_."
  if (is.null(weight)) {
    data[[w]] <- rep(1L, nrow(data))
  } else {
    stopifnot(exprs = {
      weight %in% names(data)
      !weight %in% names(c(xvars, yvars))
      !anyNA(data[[weight]])
      all(data[[weight]] >= 0)
    })
    data[[w]] <- data[[weight]] / mean(data[[weight]])  # Scaled weight to avoid numerical issues with larger integer weights
    data[[weight]] <- NULL
  }

  #-----

  # Check for character-type variables; stop with error if any detected
  x <- sapply(data[c(xvars, yvars)], is.character)
  if (any(x)) stop("Please coerce character variables to factor:\n", paste(names(which(x)), collapse = ", "))

  #-----

  # Check that the 'xvars' and 'yvars' contain only syntactically valid names
  stopifnot(all(make.names(yvars, unique = TRUE) == yvars))
  stopifnot(all(make.names(xvars, unique = TRUE) == xvars))

  # Print variable information to console
  cat(length(yvars), "fusion variables\n")
  cat(length(xvars), "initial predictor variables\n")
  cat(nrow(data), "observations\n")

  # Limit 'data' to the necessary variables
  data <- data[c(w, xvars, yvars)]

  #-----

  # Identify which of the 'yvars' are continuous
  ycont <- names(which(sapply(data[yvars], is.numeric)))

  # Observed "outer" range" (min/max) for continuous 'yvars'
  youter <- lapply(data[ycont], range, na.rm = TRUE)

  # The "inner range" for continuous 'yvars'; i.e. largest negative and smallest positive value
  # If there are no negative or positive values, 0 is returned
  yinner <- lapply(data[ycont], function(x) c(ifelse(any(x < 0), max(x[x < 0], na.rm = TRUE), 0), ifelse(any(x > 0), min(x[x > 0], na.rm = TRUE), 0)))

  # Extract data classes and levels for the 'yvars'
  x <- data[yvars]
  yclass <- lapply(x, class)
  ylevels <- lapply(x[grepl("factor", yclass)], levels)

  # Placeholder list for 'ycor' with slots for continuous and ordered factor fusion variables (but not unordered factors)
  ycor.vars <- names(which(sapply(yclass, function(x) x[1] != "factor")))

  #-----

  # Extract data classes and levels for the 'xvars'
  x <- data[xvars]
  xclass <- lapply(x, class)
  xlevels <- lapply(x[grepl("factor", xclass)], levels)

  #-----

  # Check for no-variance (constant) variables; record them in 'sim.constant' and remove from 'data'
  constant <- names(which(sapply(data[c(xvars, yvars)], novary)))
  if (length(constant)) {
    sim.constant <- map(data[intersect(constant, yvars)], ~ na.omit(unique(.x)))
    yvars <- setdiff(yvars, constant)
    xvars <- setdiff(xvars, constant)
    cat("Detected", length(constant), "constant variable(s):\n", paste(constant, collapse = ", "), "\n")
  } else {
    sim.constant <- NULL
  }

  #-----

  # Detect if unordered factors with many levels could cause problems and issue immediate warning to console
  if (is.null(maxcats) & any(map_chr(yclass, 1L) == "factor") & any(lengths(xlevels) > 15)) {
    warning("\nBe careful: Unordered factors are present that could cause long compute times.\nSee 'maxcats' argument in ?train.\n", immediate. = TRUE)
    # NOT USED: Prompt user to confirm at console that they want to continue
    #   cat("Hold up! Unordered factors are present that could cause long compute times. See 'maxcats' argument in ?train.\n")
    #   continue <- readline(prompt = "Do you want to continue with 'maxcats = NULL'? (Y/N):\n")
    #   if (tolower(continue) != "y") stop(call. = FALSE)
  }

  #-----

  # Detect and impute any missing values in 'data'
  na.cols <- names(which(sapply(data, anyNA)))
  if (length(na.cols) > 0) {
    #cat("Imputing missing values in predictor variables...\n")
    cat("Missing values imputed for the following variable(s):\n", paste(na.cols, collapse = ", "), "\n")
    for (j in na.cols) {
      x <- data[[j]]
      ind <- is.na(x)
      set(data, i = which(ind), j = j, value = imputationValue(x, ind))
    }
  }

  #-----

  # Detect 1-to-1 linear dependencies among the variables
  # This allows certain variables to be excluded from the modeling process
  # The information needed to simulate these variables in subsequent fuse() call is retained in the 'sim.lm' and 'sim.merge' lists
  cat("Searching for derivative relationships...\n")

  # Detect near-perfect bivariate relationships where the effective R-squared is above some (high) 'threshold'
  # List of slimmed 'biglm' models that can be used for simple simulation of variables identified in 'names(sim.lm)'
  sim.lm <- detectCorrelationDependence(data,
                                        fvars = yvars,
                                        exclude = grep("..", names(data), value = TRUE, fixed = TRUE),
                                        threshold = 0.99)

  # Detect perfect bivariate categorical relationships
  # List of data frames that can be used to add/merge derivative variables when variables in 'names(sim.merge)' are known
  # NOTE: 'exclude' argument below excludes fusionACS spatial predictors ("..") by default! Not particularly safe.
  #sim.merge <- detectCategoricalDependence(data, fvars = yvars, exclude = names(sim.lm))  # This is universally appropriate but slow with many fusionACS spatial predictors
  sim.merge <- detectCategoricalDependence(data,
                                           fvars = yvars,
                                           exclude = c(names(sim.lm), grep("..", names(data), value = TRUE, fixed = TRUE)),
                                           cores = cores)

  # Vector of variables identified as derivative of other variables
  derivative <- unique(c(names(sim.lm), unlist(map(sim.merge, ~ names(.x)[-1]))))

  # Variables identified as derivative can be removed from 'data'
  if (length(derivative) > 0) {
    data <- select(data, -any_of(derivative))
    cat("Detected", length(derivative), "derivative variable(s) that can be omitted from modeling\n")
  }

  #----------------------------------
  #----------------------------------

  # RANKS MATRIX

  #cat("Building ranks matrix...\n")

  # Unordered factor variables among all retained variables
  unordered <- sapply(c(xclass, yclass), function(x) x[1] == "factor")
  unordered <- unordered[names(unordered) %in% names(data)]

  # Build 'ranks' data.table for 'xvars' that are NOT unordered factors
  ranks <- subset(data, select = names(which(!unordered)))
  for (v in names(ranks)) data.table::set(ranks, j = v, value = data.table::frank(ranks[[v]], ties.method = "average"))

  # Variable associated with each column of 'ranks' matrix
  # Updated below when dummy variable columns are created
  ranks.var <- names(ranks)

  # Create dummy variable columns in 'ranks' for the 'xvars' that ARE unordered factors
  for (v in names(which(unordered))) {
    dt <- subset(data, select = v)
    u <- xlevels[[v]]
    newv <- paste0(v, u)
    ranks.var <- c(ranks.var, rep(v, length(u)))
    ranks[newv] <- lapply(u, function(x) as.integer(dt == x))  # Works if 'ranks' is data.frame or data.table
  }

  # Safety check
  stopifnot(length(ranks.var) == ncol(ranks))

  #----------------------------------
  #----------------------------------

  # data.table with zero/nonzero binary version of numeric variables
  dzero <- as.data.table(data)
  vnum <- setdiff(names(which(sapply(data, is.numeric))), w)
  for (v in vnum) set(dzero, j = v, value = dzero[[v]] == 0)

  # Make dzero unique? Would this be faster?

  #----------------------------------
  #----------------------------------

  # DETERMINE FUSION ORDER

  cat("Determining order of fusion variables...\n")

  # Adjusts 'yvars' for removal of variables that can be deduced by 1-to-1 linear relationships
  yvars <- intersect(yvars, names(data))

  # Fit full models, including yvars as predictors
  # Only 'variable.importance' slot is returned
  full.varimp <- pbapply::pblapply(X = yvars, FUN = function(y) {
    m <- fitRpart(y = y,
                  x = c(xvars, setdiff(yvars, y)),
                  w = w,
                  data = data,
                  n = initial[1] * nrow(data),
                  maxcats = maxcats,
                  linear = FALSE,
                  lasso.threshold = lasso,
                  args = list(cp = initial[2],
                              minbucket = ifelse(y %in% ycont, node.obs[1], node.obs[2]),
                              xval = 0,
                              maxcompete = 0)
    )
  }, cl = cores) %>%
    setNames(yvars)

  # Find preferred order of yvars
  yord <- fusionOrder(varimp = map(full.varimp, "variable.importance"))

  # Cleanup
  rm(full.varimp)
  gc()

  #----------------------------------
  #----------------------------------

  # FIT MODELS

  cat("Building fusion models...\n")

  # Function to fit rpart() model and calculate rank correlations for i-th response variable in 'yord'
  fitModel <- function(i) {

    y <- yord[i]

    # Is 'y' continuous?
    cont <- y %in% ycont

    # Prior response variables in the fusion sequence
    yprior <- if (i == 1) NULL else yord[1:(i - 1)]

    #--------

    # Restrict data to rows with non-NA response values
    ind <- which(!is.na(data[[y]]))
    d <- data[ind, c(w, y, yprior, xvars)]

    #--------

    # Detect if there is a numeric predictor with zero/nonzero structure that is derivative of 'y'
    # If a zero/nonzero binary version of a predictor can explain 'y', then we want to include that binary version and force its inclusion
    # Example: one of the predictors is natural gas consumption (numeric) and 'y' is categorical indicating if household uses natural gas
    xbin <- catDependence(data = dzero,
                          targets = y,
                          candidates = intersect(vnum, c(xvars, yprior)),
                          cores = 1L)

    # Update 'd' with binary version of 'xbin' derivative variables
    # In such cases, the binary variable generally
    if (is.null(xbin)) {
      xbinary <- NULL
    } else {
      xbin <- xbin[[1]][1]  # Retains only a single explanator variable, even if multiple exist
      xbinary <- paste0(xbin, "_BINARY_")
      d <- cbind(d, setnames(x = dzero[, ..xbin], new = xbinary))

      # Extract any unique 'y' values associated with binary predictor
      xmerge <- d %>%
        select(all_of(c(xbinary, y))) %>%
        distinct() %>%
        group_by_at(xbinary) %>%
        filter(n() == 1) %>%
        ungroup()

      # Subset 'd' so that decision tree only fit to subset of values not explained by 'xmerge'
      d <- subset(d, !d[[y]] %in% xmerge[[y]])

    }

    #--------

    # Fit prediction model
    m <- fitRpart(y = y,
                  x = c(xvars, yprior, xbinary),
                  w = w,
                  data = d,
                  maxcats = maxcats,
                  n = nrow(d),
                  linear = TRUE,
                  lasso.threshold = lasso,
                  args = list(cp = complexity,
                              minbucket = ifelse(y %in% ycont, node.obs[1], node.obs[2]),
                              xval = 0,
                              maxcompete = 0)
    )

    #--------

    # Add 'xbin' slot indicating which numeric variables where binarized prior to model-fitting
    # Add 'xmerge' slot indicating which binary numeric predictors perfectly explain an outcome value
    # If the response variable is categorical, remove the "known" level from the 'ylevels' attribute of the rpart object (causes prediction errors otherwise)
    if (!is.null(xbinary)) {
      m$xbin <- xbin
      m$xmerge <- xmerge
      #if (class(m) == "rpart" & ((is.factor(xmerge[[2]] & !is.ordered(xmerge[[2]])) | is.logical(xmerge[[2]]))) {  # Not clear if is.logical condition is correct here! Not thoroughly tested
      if (class(m) == "rpart" & (is.factor(xmerge[[2]]) & !is.ordered(xmerge[[2]]))) {  # Omitting is.logical condition until clear it is necessary
        attr(m, "ylevels") <- setdiff(attr(m, "ylevels"), xmerge[[2]])
      }
    }

    #--------

    # If 'y' is continuous...
    if (cont) {

      # Add additional slots to rpart model output
      if (class(m) == "rpart") {

        # Create list in 'm' to hold the quantile values associated with a set of 'N' known percentiles; see fitDensity()
        # The 'Q' values describe the shape of the quantile function (i.e. inverse CDF)
        nodes <- sort(unique(m$where))

        # Placeholder matrix for the smoothed quantile values
        Qn <- 500  # Ideally, would be a train() control argument
        m$Q <- matrix(0L, nrow = Qn, ncol = length(nodes))
        colnames(m$Q) <- nodes

        # Fit density to observations in each node
        for (n in nodes) {

          # Index identifying observations in node 'n'
          nind <- m$where == n

          # Node density result
          fd <- fitDensity(x = d[nind, y], w = d[nind, w], inner.range = yinner[[y]], outer.range = youter[[y]], N = Qn)

          # Add quantile function values to 'm'
          m$Q[, as.character(n)] <- fd

        }

      }

    }

    #--------

    # Calculate (rank) correlation between 'y' and existing variables in 'ranks' (if necessary; only continuous and ordered factors)
    if (y %in% ycor.vars) {
      #ind <- which(as.numeric(data[[y]]) != 0)  # Restrict correlation calculation to non-zero observations
      ind <- 1:nrow(data)  # No restriction on correlation calculation
      j <- ranks.var == y  # Column in 'ranks' associated with 'y'
      k <- ranks.var %in% c(xvars, yprior)  # Columns in 'ranks' associated with predictors of 'y'
      p <- suppressWarnings(cor(ranks[ind, k], ranks[ind, j]))[, 1]  # Will silently return NA if no variance in a particular column
      p <- cleanNumeric(p[!is.na(p)], tol = 0.001)
    } else {
      p <- NULL
    }

    # Cleanup
    gc()  # Perhaps useful when run in parallel

    return(list(m = slimRpart(m), p = p))

  }

  #-----

  # Call fitModel() for each 'yord', possibly in parallel
  mout <- pbapply::pblapply(
    X = 1:length(yord),
    FUN = fitModel,
    cl = cores) %>%
    setNames(yord)

  # test <- sapply(mout, class)
  # which(test == "try-error")
  # check <- lapply(which(test == "try-error"), fitModel)

  # Troubleshoot
  #for (i in 1:length(yord)) fitModel(i)

  # Check certain outputs
  # map(mout, ~ .x$m$xbin)
  # table(map_chr(mout, ~class(.x$m)))

  #-----

  # Assemble final result
  result <- list(
    models = map(mout, "m"),
    xclass = xclass,
    xlevels = xlevels,
    yclass = yclass,
    ylevels = ylevels,
    yinner = yinner,
    youter = youter,
    ycor = compact(map(mout, "p")),
    derivative = list(constant = sim.constant,
                      model = sim.lm,
                      merge = sim.merge)
  )

  return(result)

}
